File: javascript/sdk/caas.js
/**
 * CAAS module - Direct Debit (Charging as a Service)
 * Equivalent to PHP DirectDebitSender class in sdk_file.php
 */

const Core = require('./core');

// ─── Custom Exception ───────────────────────────────────────────────
class CassException extends Error {
  constructor(message, code, response = null) {
    super(message);
    this.name = 'CassException';
    this.statusMessage = message;
    this.code = code;
    this.rawResponse = response;
  }

  getStatusCode() { return this.code; }
  getStatusMessage() { return this.statusMessage; }
  getRawResponse() { return this.rawResponse; }
}

// ─── Direct Debit Sender ────────────────────────────────────────────
class DirectDebitSender extends Core {
  /**
   * @param {string} serverURL - BDApps CAAS direct debit URL
   * @param {string} applicationId - Application ID
   * @param {string} password - Application password
   */
  constructor(serverURL, applicationId, password) {
    super();
    this.serverURL = serverURL;
    this.applicationId = applicationId;
    this.password = password;
  }

  /**
   * Charge a subscriber via direct debit
   * @param {string} externalTrxId - External transaction ID
   * @param {string} subscriberId - Subscriber phone number (tel:880...)
   * @param {number|string} amount - Amount to charge
   * @returns {Promise<string>} Status code 'S1000' on success
   */
  async cass(externalTrxId, subscriberId, amount) {
    if (!subscriberId) {
      throw new Error('Address should be a string or an array of strings');
    }
    return this._cassMany(externalTrxId, subscriberId, amount);
  }

  /** @private */
  async _cassMany(externalTrxId, subscriberId, amount) {
    const payload = {
      applicationId: this.applicationId,
      password: this.password,
      externalTrxId,
      subscriberId,
      paymentInstrumentName: 'Mobile Account',
      amount: String(amount),
    };

    const response = await this.sendRequest(payload, this.serverURL);
    return this._handleResponse(response);
  }

  /** @private */
  _handleResponse(jsonResponse) {
    if (!jsonResponse) {
      throw new CassException('Invalid server URL', '500');
    }

    const { statusCode, statusDetail } = jsonResponse;

    if (statusCode === 'S1000') {
      return 'S1000';
    }

    throw new CassException(statusDetail || 'Unknown error', statusCode || 'UNKNOWN');
  }
}

// ─── Balance Query ──────────────────────────────────────────────────
class BalanceQuery extends Core {
  constructor(applicationId, password) {
    super();
    this.applicationId = applicationId;
    this.password = password;
    this.serverURL = 'https://developer.bdapps.com/caas/balance/query';
  }

  /**
   * Query the balance of a subscriber
   * @param {string} subscriberId - Subscriber phone number
   * @param {string} paymentInstrumentName - Payment instrument (default: 'Mobile Account')
   * @returns {Promise<object>} Balance info
   */
  async queryBalance(subscriberId, paymentInstrumentName = 'Mobile Account') {
    const payload = {
      applicationId: this.applicationId,
      password: this.password,
      subscriberId,
      paymentInstrumentName,
    };

    return this.sendRequest(payload, this.serverURL);
  }
}

// ─── Payment Instrument List ────────────────────────────────────────
class PaymentInstrumentList extends Core {
  constructor(applicationId, password) {
    super();
    this.applicationId = applicationId;
    this.password = password;
    this.serverURL = 'https://developer.bdapps.com/caas/list/pi';
  }

  /**
   * Get payment instruments for a subscriber
   * @param {string} subscriberId - Subscriber phone number
   * @param {string} type - Type filter: 'sync', 'async', or 'all'
   * @returns {Promise<object>} Payment instrument list
   */
  async getList(subscriberId, type = 'all') {
    const payload = {
      applicationId: this.applicationId,
      password: this.password,
      subscriberId,
      type,
    };

    return this.sendRequest(payload, this.serverURL);
  }
}

module.exports = { DirectDebitSender, BalanceQuery, PaymentInstrumentList, CassException };

File: python/sdk/caas.py
"""
CAAS module - Direct Debit / Balance Query / Payment Instruments.
Equivalent to PHP DirectDebitSender class in sdk_file.php.
"""

from typing import Optional, Union

from .core import Core
from .exceptions import BDAppsException


class CassException(BDAppsException):
    """Exception for CAAS service errors."""
    pass


class DirectDebitSender(Core):
    """
    Charge a subscriber via direct debit.
    Equivalent to PHP DirectDebitSender class.
    """

    def __init__(self, server_url: str, application_id: str, password: str):
        super().__init__()
        self.server_url = server_url
        self.application_id = application_id
        self.password = password

    async def cass(
        self, external_trx_id: str, subscriber_id: str, amount: Union[str, int, float]
    ) -> str:
        """
        Charge a subscriber via direct debit.

        Args:
            external_trx_id: External transaction ID.
            subscriber_id: Subscriber phone number (tel:880...).
            amount: Amount to charge.

        Returns:
            Status code 'S1000' on success.
        """
        if not subscriber_id:
            raise ValueError("Address should be a string or an array of strings")

        return await self._cass_many(external_trx_id, subscriber_id, amount)

    async def _cass_many(
        self, external_trx_id: str, subscriber_id: str, amount: Union[str, int, float]
    ) -> str:
        payload = {
            "applicationId": self.application_id,
            "password": self.password,
            "externalTrxId": external_trx_id,
            "subscriberId": subscriber_id,
            "paymentInstrumentName": "Mobile Account",
            "amount": str(amount),
        }

        response = await self.send_request(payload, self.server_url)
        return self._handle_response(response)

    def _handle_response(self, json_response: dict) -> str:
        if not json_response:
            raise CassException("Invalid server URL", "500")

        status_code = json_response.get("statusCode", "")
        status_detail = json_response.get("statusDetail", "Unknown error")

        if status_code == "S1000":
            return "S1000"

        raise CassException(status_detail, status_code)


class BalanceQuery(Core):
    """Query subscriber account balance."""

    def __init__(self, application_id: str, password: str):
        super().__init__()
        self.application_id = application_id
        self.password = password
        self.server_url = "https://developer.bdapps.com/caas/balance/query"

    async def query_balance(
        self,
        subscriber_id: str,
        payment_instrument_name: str = "Mobile Account",
    ) -> dict:
        """
        Query the balance of a subscriber.

        Args:
            subscriber_id: Subscriber phone number.
            payment_instrument_name: Payment instrument (default: 'Mobile Account').

        Returns:
            Balance info dictionary.
        """
        payload = {
            "applicationId": self.application_id,
            "password": self.password,
            "subscriberId": subscriber_id,
            "paymentInstrumentName": payment_instrument_name,
        }

        return await self.send_request(payload, self.server_url)


class PaymentInstrumentList(Core):
    """Get available payment instruments for a subscriber."""

    def __init__(self, application_id: str, password: str):
        super().__init__()
        self.application_id = application_id
        self.password = password
        self.server_url = "https://developer.bdapps.com/caas/list/pi"

    async def get_list(
        self, subscriber_id: str, pi_type: str = "all"
    ) -> dict:
        """
        Get payment instruments for a subscriber.

        Args:
            subscriber_id: Subscriber phone number.
            pi_type: Type filter: 'sync', 'async', or 'all'.

        Returns:
            Payment instrument list dictionary.
        """
        payload = {
            "applicationId": self.application_id,
            "password": self.password,
            "subscriberId": subscriber_id,
            "type": pi_type,
        }

        return await self.send_request(payload, self.server_url)
